### ISSUE TITLE
QuakeML verification problem (related to SC3ML converted using stylesheets)

### ISSUE THREAD

**Issue by megies** [obspy/obspy#1872]
I think there is a problem with the QuakeML RelaxNG schema (upstream). It seems to fail validation on URIs that do not contradict the regular expression pattern specified in the QuakeML documentation.

Here's a minimal example QuakeML:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<q:quakeml xmlns:q="http://quakeml.org/xmlns/quakeml/1.2" xmlns="http://quakeml.org/xmlns/bed/1.2">
  <eventParameters publicID="smi:org.gfz-potsdam.de/geofon/NA">
    <event publicID="smi:org.gfz-potsdam.de/geofon/gof2017qznj">
      <stationMagnitude publicID="smi:org.gfz-potsdam.de/geofon/Origin#20170830154555.615079.1673910#staMag.MLv#GR.WET">
        <originID>smi:org.gfz-potsdam.de/geofon/Origin#20170830154555.615079.1673910</originID>
        <mag>
          <value>1.597395494</value>
        </mag>
        <type>MLv</type>
        <amplitudeID>smi:org.gfz-potsdam.de/geofon/Amplitude#20170830154537.481105.2138874</amplitudeID>
        <waveformID networkCode="GR" stationCode="WET" channelCode="HHZ"/>
        <creationInfo>
          <agencyID>GOF</agencyID>
          <author>scmag@kelut</author>
          <creationTime>2017-08-30T15:45:55.803111Z</creationTime>
        </creationInfo>
      </stationMagnitude>
    </event>
  </eventParameters>
</q:quakeml>
```

Validation against the RelaxNG schema fails:
```
$ jing ~/git/obspy/obspy/io/quakeml/data/QuakeML-1.2.rng quakeml.xml
/tmp/xslt/quakeml.xml:5:121: error: value of attribute "publicID" is invalid; must be a URI
```

The problem are the additional `#` characters in the local identifier part (which itself is denoted by the first `#` character). If these are removed, then the file validates just fine. Also, if the RelaxNG schema is changed from type `anyURI` to `string` for the `ResourceIdentifier` type (that includes the regex pattern for verification anyway), than it also verifies OK. So the problem seems that the `anyURI` is adding additional constraints and disallows multiple `#` characters.

However, the regex pattern specified by QuakeML documentation (and in the RelaxNG schema) clearly allows these additional `#` characters within the local identifier field (compare the last group of the regex: `[\w\d\-\.\*\(\)\+\?_~'=,;#/&amp;]*`):

```
<rng:param name="pattern">(smi|quakeml):[\w\d][\w\d\-\.\*\(\)_~']{2,}/[\w\d\-\.\*\(\)_~'][\w\d\-\.\*\(\)\+\?_~'=,;#/&amp;]*</rng:param>
```

The QuakeML official documentation pdf in the text however suggests that additional `#` characters are not allowed:

```
The resource-id may be followed by a stop character ("#") and a local identifier which
can be made up of alphanumeric characters, the comma (",") and semicolon (";") characters,
and the characters from the following list: -, _, ., ~, *, ´, (, ), /, +, =, ?. 
```

---

Long story short:

This ambiguity in whether multiple `#` characters are allowed or not is causing validation problems and also affects documents converted using SC3ML -> QuakeML stylesheet (`obspy.io.seiscomp` read routines -- the above minimal example was extracted from a converted SC3ML document, seiscomp seems to generate publicIDs like the following in SC3ML: `<stationMagnitude publicID="Origin#20170830154555.615079.1673910#staMag.MLv#GR.WET">`).

And I guess we need to get a statement from QuakeML guys what exactly is valid now.. multiple hash signs allowed or not.

CC @Brtle @Jollyfant @gempa-stephan

**Comment by Brtle** [comment]
I had exactly the same issue today. As you said, it probably comes from the [anyURI](http://www.datypic.com/sc/xsd/t-xsd_anyURI.html) which does not allow several `#` character. Perhaps we can just replace the `#` character with a `_` character during the conversion?

**Comment by megies** [comment]
Ideally this should be handled upstream at http://geofon.gfz-potsdam.de/schema. In general it's best to have as few local modifications to these schema as possible and have most fixes merged upstream.. to ease up maintenance in the future.

But also the QuakeML guys should be contacted because there's clearly an ambiguity in the QuakeML documentation.

**Comment by Brtle** [comment]
I did some changes to the XSLT files to keep only the first `#` character and replace the others by `_`. I think it should be better to have a SC3ML module which generate valid QuakeML file for the 1.1.0.

By the way, QuakeML documentation and validation files (both XSD and RelaxNG) are consistent since they both forbid multiple `#` characters event though it's not very clear with the regular expression.

If this is suitable for everyone, can you also upgrade the upstream XSLT files @gempa-stephan?

**Comment by gempa-stephan** [comment]
I think it is not a good idea to convert ``#`` into ``_`` just to please the QuakeML validator.

When converting back from QuakeML to SC3ML you can't decided whether the ``_`` needs to be converted to ``#`` or should remain unchanged. Please note that the public ID pattern is configurable in SeisComP3. So every agency may use their own pattern and characters.

IDs should not have a semantic. There should be no code which splits the ID into parts separated by ``#`` and interprets the parts for whatever reason. I think a change request should be filed against QuakeML which asks for a more relax ID definition.

**Comment by gempa-jabe** [comment]
FYI, I have pushed a change to SC3 that removes hashes from the default publicID pattern. Anyway, I always found the QuakeML publicID constraints too restrictive and a little ridiculous but it is in the standard, nobody complained so far and people start to use it due to the web services. I am not in favour of converting an ID because as the name suggests, it is an ID which is used to identify objects. If you rename them (possibly with different converters and different rules), the global uniqueness of an object is not given anymore. That said, I can't change the QuakeML state right now but I am writing all that to make you aware of the issue (although I am sure that you are already) and possibly look in more detail at the next revision of the RFC. I will ;)

For now I could imagine to use URL encoding for publicIDs. Not sure right now if that encoding is part of the anyURI specs. According to the specs, it discourages the use of spaces unless they are encoded by %20. So URL encoding should be fine.

**Comment by megies** [comment]
I agree that replacing characters is potentially disruptive of the whole ID idea. @Brtle would you mind contacting the QuakeML guys to ask if the schema can be changed to allow multiple hash characters?

In any case, this is once more showing that everybody should actually honor the available schema documents whenever creating QuakeML (or other schema-ed documents). After all this is really a problem of documents out there that don't match the provided schema...

Thanks @gempa-jabe for removing use of multiple hash characters in the default seiscomp ID naming scheme. That's the easiest solution, at least for future documents..

**Comment by Brtle** [comment]
> I agree that replacing characters is potentially disruptive of the whole ID idea. @Brtle would you mind contacting the QuakeML guys to ask if the schema can be changed to allow multiple hash characters?

Done, I sent an email to the QuakeML guys.

I will close my PR since I was going in the wrong direction. Thank you for your feedback!

**Comment by Brtle** [comment]
I got a response from the QuakeML team:

> QuakeML publicIDs are Uniform Ressource Identifiers (URIs) following the RFC2396 of the Internet Engineering Task Force (Berners-Lee 1998, see http://www.ietf.org/rfc/rfc2396.txt)
They follow a generic pattern [smi|quakeml]: authority-id/resource-id[#local-id]  *
In this standard, the # sign is reserved to separate the globally unique id from an optional extension with significance only for the ID provider. For details, see chapter 3.1 of the QuakeML BED 1.2 standard document (https://quake.ethz.ch/quakeml/docs/REC?action=AttachFile&do=view&target=QuakeML-BED-20130214a.pdf)
>
> The SeisComP 3 software does not enforce  publicIDs of its database or SeisCompML (which is similar, but not identical to quakeml) to be globally unique (e.g. by using an installation-specific prefix), nor to be compliant with the QuakeML standard (in contrary, default configurations use # as separator of publicID fragments, resulting to multiple # characters in ID strings as you observe it.
However, SeisComP3 allows  to configure the publicID pattern. Thus, it is in the responsibility of the operator to configure for compatibility with QuakeML.
>
> This can be done e.g by setting the property publicIDPattern in global.cfg to:
>
> publicIDPattern="smi:org.institution.subunit/seiscompSystemId/@classname@/@time/%Y%m%d%H%M%S.%f%.@id@"
>
> where 
> - org.institution.subunit and seiscompSystemId stand for strings of your choice. For the first of them, it may be an idea to use the inverse of your web domain, as the DNS system already ensures that this prefix is globally unique. For seiscompSystemID, you have to make sure that you use a different string for each seiscomp instance you have
> - the @bla@ variables refer to properties of the element identified, with @id@ being its seiscomp-internal id
> Like this, the resulting publicID is globally unique AND quakeML compatible. For seiscomp-internal variables available in these patterns, see: 
> 
> you can define different patterns for different modules  by redeclaring the pattern in the module's .cfg file.
> Some modules have also separate patterns for specific purposes, for example the eventIDPattern in scevent.cfg 
>
>
> Best regards, 
>
> Philipp
> 
> ----------------
> *for QuakeML 2.0, we plan to relax the scheme requirement from smi or quakeml to actually any URI scheme, which would allow to use also http, DOI, or anything you might, or might not find in the IANA scheme registry (https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml). However, the generic concept of using an URI for global identification proved quite useful in the last 20 years and should not be challenged.

The last sentence is quite clear. I think we can expect no change in the QuakeML schema.

**Comment by megies** [comment]
So, I really don't know how to handle this.. modifying the existing ID would break potential links to identified objects when converting back and forth, as also @gempa-jabe pointed out. On the other hand, simply adapting our rng schema to allow multiple hashes would mean we're silently producing invalid QuakeML, which is also not really desireable..

How about a way in the middle? We could relax the ID check, and write the IDs unmodified and in the process show a warning that notifies of the schema-breaching IDs with a notice that the ID naming scheme should be changed by the author of the documents?

CC @krischer

**Comment by Brtle** [comment]
Does the ObsPy ResourceIdentifier allow these multiple # characters? Currently the SC3ML read module can generate unvalid QuakeML files but working ObsPy catalogs. Maybe we can just handle a warning message in the QuakeML write module.

**Comment by megies** [comment]
The regex we have in the ResourceIdentifier is the regex from the QuakeML manual text part, which allows multiple hashes, I think.. see 
 - https://github.com/obspy/obspy/blob/7d8f8aea72d3c81ade4cc17036ef2be97cacce0a/obspy/core/event/base.py#L438-L441
 - https://github.com/obspy/obspy/blob/7d8f8aea72d3c81ade4cc17036ef2be97cacce0a/obspy/core/event/base.py#L782-L784

**Comment by krischer** [comment]
> How about a way in the middle? We could relax the ID check, and write the IDs unmodified and in the process show a warning that notifies of the schema-breaching IDs with a notice that the ID naming scheme should be changed by the author of the documents?
> 
> CC @krischer

Sounds good to me. Our current `ResourceIdentifier` object work fine with multiple hashes. While actually writing the QuakeML we can just check with the stricter QuakeML schema regex and raise a warning so people are aware of something not going according to plan.

The only problem with this are the then slightly misnamed `ResourceIdentifier.convert_id_to_quakeml_uri()` and `ResourceIdentifier().get_quakeml_uri()` methods which (as they also do now) would happily return ids with multiple hashes as we used the regex from the QuakeML documentation.

The upside of this whole discussion is that our internal event objects will be a bit less dependent on QuakeML as a format which is IMHO a desirable thing.

**Comment by megies** [comment]
> The only problem with this are the then slightly misnamed ResourceIdentifier.convert_id_to_quakeml_uri() and ResourceIdentifier().get_quakeml_uri() methods which (as they also do now) would happily return ids with multiple hashes as we used the regex from the QuakeML documentation.

Maybe put the warning inside those routines then? We could also add a `loose=True` kwarg (to those and to `_write_quakeml`) so that the behavior is cleanly documented?

**Comment by claudiodsf** [comment]
Just for my reference, the change in SC3 mentioned by @gempa-jabe should be this one: https://github.com/SeisComP3/seiscomp3/commit/e71e904332fbb12b4b93be37106759ebce36e8be

**Comment by claudiodsf** [comment]
I would just close it, since the non-conforming behavior has been fixed on SC3 side.
